orttraining/orttraining/core/optimizer/transformer_layer_recompute.cc

```diff
@@ -0,0 +1,180 @@
+        layer_end_edges.push_back(next_node->OutputDefs()[0]);
+      }
+    }
+  }
+
+  ORT_RETURN_IF_NOT(layer_start_edges.size() == layer_end_edges.size(), "Number of start and end edges doesn't match!");
+
+  start_end_edges.clear();
+
+  LOGS(logger, INFO) << "Found " << layer_start_edges.size() << " transformer layers.";
+  for (size_t i = 0; i < layer_start_edges.size(); ++i) {
+    start_end_edges.push_back({layer_start_edges[i], layer_end_edges[i]});
+    LOGS(logger, INFO) << "Start edge: " << layer_start_edges[i]->Name() << " End edge: " << layer_end_edges[i]->Name();
+  }
+
+  return Status::OK();
+}
+
+namespace {
+
+typedef std::set<const Node*, NodeCompare> NodeSet;
+
+NodeSet BFSFrom(const std::vector<const Node*>& start_nodes, bool reverse) {
+  NodeSet visited(start_nodes.begin(), start_nodes.end());
+  std::deque<const Node*> queue(start_nodes.begin(), start_nodes.end());
+  while (!queue.empty()) {
+    const Node* n = queue.front();
+    queue.pop_front();
+
+    auto begin = reverse ? n->InputNodesBegin() : n->OutputNodesBegin();
+    auto end = reverse ? n->InputNodesEnd() : n->OutputNodesEnd();
+
+    for (auto node_it = begin; node_it != end; ++node_it) {
+      const Node& node = *node_it;
+      if (visited.find(&node) == visited.end()) {
+        queue.push_back(&node);
+        visited.insert(&node);
+      }
+    }
+  }
+  return visited;
+}
+}  // namespace
+
+std::vector<const Node*> TransformerLayerRecompute::NodesBetweenEdges(const Graph& graph, const NodeArg* start, const NodeArg* end) const {
+  // Forward BFS from the start node
+  std::vector<const Node*> start_nodes = graph.GetConsumerNodes(start->Name());
+  NodeSet fw_visited = BFSFrom(start_nodes, /*reverse*/ false);
+
+  // Reverse BFS from the end node
+  const Node* end_node = graph.GetProducerNode(end->Name());
+  NodeSet bw_visited = BFSFrom({end_node}, /*reverse*/ true);
+
+  // Join fw_visited and bw_visited
+  std::vector<const Node*> intersect_nodes;
+  std::set_intersection(fw_visited.begin(), fw_visited.end(),
+                        bw_visited.begin(), bw_visited.end(),
+                        std::back_inserter(intersect_nodes), NodeCompare());
+
+  return intersect_nodes;
+}
+
+void TransformerLayerRecompute::InsertRecomputeNodes(Graph& graph, const std::vector<const Node*>& nodes, int priority) const {
+  auto initializers = graph.GetAllInitializedTensors();
+
+  for (const Node* n : nodes) {
+    Node* node = graph.GetNode(n->Index());
+
+    if (node->OpType() == "Dropout") {
+      const NodeArg* input = node->InputDefs()[0];
+      const Node* p_node = graph.GetProducerNode(input->Name());
+
+      bool use_original_input =
+          initializers.find(input->Name()) != initializers.end() ||
+          std::find(nodes.begin(), nodes.end(), p_node) == nodes.end();
+
+      Node& recompute_node = InsertDropoutRecompute(graph, *node, use_original_input);
+      recompute_node.SetPriority(priority);
+      continue;
+    }
```

Comment:
Why Dropout need a special case? Some comments might help me understanding

Reply:
Recomputed dropout need to replicate the dropout pattern in the original node. 
Comment added.